### ISSUE TITLE
New build strategy to trigger external CI (Jenkins)

### ISSUE THREAD

**Issue by jimmidyson** [openshift/origin#6954]
In order to support things like the fabric8 CD pipelines, we'd like to add a new build strategy that triggers external CI via a webhook. This build strategy would be configured just with a URL & optionally a secret to use to trigger the build in an external CI server.

This would also involve adding the creation of a Jenkins job via `oc new-app` if the source repository contains the standard `Jenkinsfile` used for pipeline configuration.

/cc @jstrachan @rawlingsj

**Comment by bparees** [comment]
@smarterclayton has opinions on this.

**Comment by smarterclayton** [comment]
I'm fine with a "Jenkins" strategy if we can figure out how to appropriately scope it.  However, when I create this build config, what happens if Jenkins isn't installed?  How does OpenShift securely communicate with Jenkins (web hook would need auth, where is that auth stored)?  How does build status get written back to OpenShift (will Jenkins have auth to mark OpenShift builds as complete and record which job triggered it)?  Does a Jenkins have to be installed in the current project with a special name in order to be able to create a build config with this strategy?  How does a user get notified that they need to have their project admin install and configure the jenkins integration?

If this was a "watching" pattern vs a "web hook" pattern, Jenkins would watch for new builds and then update their status as jenkins runs them.  That would remove some of these problems.  It might not even be necessary to allow regular "Build" objects to be created - instead, Jenkins would take the build config configuration and write back status to OpenShift on the build config itself to indicate the progress of jobs, and the user would click through into jenkins.  Would a user expect `oc start-build` to do something if they had jerkins installed?  Or would it be better to send the user to jenkins, because jenkins is going to have the UI around creating the build correctly (esp. if we have single sign on?).

A bunch of questions, but spitballing:

1. User creates a build config of type "Jenkins" or "Pipeline" or "MagicUnicorns"
2. OpenShift has something that tells the user whether Jenkins is installed or not in the build config status ("You've created a pipeline build, but you need to install the Jenkins integration in this project") and in the future we might have auto create for you
3. Jenkins starts up, uses its service account token to scan the build configs and create build jobs for them, then writes status to each build config "build config is run by jenkins, job url is https://....."
4. User who invokes start-build gets told "go to this URL in your browser"
5. Jenkins does whatever else it needs directly via the UI

?

**Comment by mfojtik** [comment]
@smarterclayton the current `openshift/jenkins-1-centos7` image is valid S2I builder image, which means you can point a S2I BuildConfig to your custom repository with your Jenkins configuration, jobs, plugins, .s2i/bin scripts and build fully customized Jenkins image.

By doing that you can also override "run" script and have the logic of auto-discovery other BuildConfig(s) and creating Jenkins jobs for them every time the Jenkins is started. In fact we already did this when we were discovering ImageStreams with "jenkins-slave" label and converting them into configuration for the kubernetes jenkins plugin.

If we have ability in OpenShift to mark certain builds that they are trigger by Jenkins (another Build trigger?), then we can present this in web console and in CLI. IOW. you start that build and you get back URL to Jenkins to watch it.

**Comment by mfojtik** [comment]
Having some of this magic happen automatically, we can just create "openshift/jenkins-builder" that will base on top of official Jenkins image and just add the "run" script logic for all this. The "run" can even be a Go static binary which will match with other builders we have.

**Comment by smarterclayton** [comment]
That seems useful, but lower level.  I really, really, really REALLY do not
like triggers - either as an implementation tool or as an architectural
pattern.  Triggers don't guarantee delivery and cannot handle the vagaries
of the distributed system.  I want declarative state of the behavior...
i.e.:

1. User "declares" they want a jenkins build with certain configuration
(source URI, base image, jenkins workflow customizations, etc)
2. A component observes that they need a jenkins builder in order to do that
3. A jenkins builder ensures a job exists to cover it.
4. The jenkins builder is responsible for ensuring that the builder is
appropriately updated.

What you're describing is certainly a different way of doing declarative -
the s2i base image for jenkins accepts a source code repository that is a
declarative definition of a jenkins workflow.  However, the integrations
we'll need for pipelines and workflows are probably going to demands humans
to get involved and interact with the system to alter state - .travis.yml
is cool, but one reason people use Jenkins, and it's to have an easy way of
evolving / iterating on your flows.  If we can easily allow both sides (job
builders / workflow tools capture definitions in an easy way and add those
back into the jenkins source definitions), all the better.  But if Jenkins
is first class, it needs the other operational patterns that we make
mandatory everywhere else.  Does every Jenkins require a PV?  What happens
when the masters go down or auth gets refreshed - is Jenkins actively
attempting to ensure it has a consistent view of the world?  How does an
operations team migrate 10k jenkins instances to a new version safely and
rollback on failure?  Images and an image pattern is useful, but I don't
think it's enough.

**Comment by jimmidyson** [comment]
Just to clarify, you see Jenkins integration working like this:

1. A new build strategy called e.g. `Jenkins`. This would be used for declaring requested state, only used for storage.
2. A Jenkins plugin watches OpenShift build configs for Jenkins build strategies, creates/updates jobs as required.
3. Part of Jenkins workflow is to update OpenShift builds/build configs with status, URLs, etc.

So basically use OpenShift to store required job configuration & Jenkins to perform builds, create build pods, etc & push results back to OpenShift.

Seems best for us to leave build trigger integration to Jenkins as well.

**Comment by jstrachan** [comment]
@smarterclayton if the CIBuilder can't communicate with the Jenkins / Teamcity / Bamboo server it can fail the build; just like if a git clone fails. 

When doing something like `oc new-app` with a git repo that has a Jenkinsfile inside; we can check if there is a service called `jenkins` in the current namespace and if not the command can fail and tell the user to configure a jenkins service (via running a jenkins template or service linking to a jenkins that is running elsewhere)

**Comment by jimmidyson** [comment]
@jstrachan Like the idea for `oc new-app`. Can validate if `Jenkinsfile` exists in source repo if `jenkins` build strategy is requested via cmd line flag. If strategy isn't specified & `Jenkinsfile` exists in source repo then can recommend `jenkins` build strategy. Auto-provisioning Jenkins as required would be good too.

When you say CI builder what do you mean? If we just use OpenShift for storage of the build config & Jenkins to run the actual builds then I don't see where the CI builder fits? But very possibly missing something.

**Comment by mfojtik** [comment]
@jstrachan @jimmidyson is the `Jenkinsfile` something widely used by the Jenkins community? I like @jstrachan idea.

**Comment by rawlingsj** [comment]
@mfojtik yeah Jenkinsfile is the main effort behind [Jenkins 2.0 - Pipeline as code](https://wiki.jenkins-ci.org/display/JENKINS/2.0+Pipeline+as+Code)

**Comment by jstrachan** [comment]
@jimmidyson my use of the term "CIBuilder" - I just meant this new kind of BuildConfig kind. Sorry for being unclear.

I guess we need a jenkins bridge microservice; that watches BuildConfigs to create/update Jenkins jobs - then also watches Jenkins builds to ensure that the Build objects get created and updated with the status of the build etc. e.g. if a build gets triggered in the Jenkins console; it'd be nice for consistency for there always to be a Build object in OpenShift etc.

Initially I figured we'd need a docker image to trigger the jenkins job when someone does an `oc start-build` - which then tails the log and passes or fails when the job completes. Though the issue is what if a jenkins build is triggered outside of OpenShift; we still may need a Build to be created. So maybe the jenkins bridge looks for jobs that don't have a Build and fills in the blanks?

**Comment by mfojtik** [comment]
CC @arilivigni

**Comment by jstrachan** [comment]
@mfojtik BTW Jenkins can now be pointed at a github organisation; it then scans all projects and for any branch that has a Jenkinsfile inside it in the root directly; it automatically spins up a Jenkins job for running the Jenkinsfile pipeline

**Comment by rawlingsj** [comment]
@jstrachan yeah I think a jenkins bridge as you describe makes sense.

**Comment by jstrachan** [comment]
I wonder should be be Jenkins specific right now or design up front for any CI tool from the start?

e.g. if a customer has an existing Teamcity or Bamboo installation and set of build jobs already (which seems fairly common right now), it would be nice to have a "TeamCity bridge" or "Bamboo bridge" that does the same kinda thing as a Jenkins bridge (watches BuildConfigs & ensures there's builds defined in TeamCity/Bamboo and keeps Builds up to date for builds triggered directly inside TeamCity/Bamboo which have an associated BuildConfig).

**Comment by arilivigni** [comment]
@jstrachan Does the Jenkinsfile and workflow 2.0 support Team City and Bamboo?

**Comment by jstrachan** [comment]
@smarterclayton on the operational side of jenkins; the jenkins could be hosted elsewhere, it could be a SaaS offering, or it could be a service and pod inside OpenShift. 

I guess we could maybe use the new Service Catalog / ServiceClaim stuff to ensure that a suitable Jenkins service implementation gets provisioned for a users namespace? i.e. if there's no Service yet for jenkins, we create a Service and ServiceClaim to trigger the provisioning of Jenkins?

Initially I was thinking we'd just have the actual URL of the build job in Jenkins to trigger inside the BuildConfig; now I'm wondering if it should be a relative URL and we use a service name - to help force the use of a CI Service / ServiceClaim?

**Comment by jstrachan** [comment]
@arilivigni a Jenkins pipeline can trigger builds in any of TeamCity, Bamboo, Jenkins or OpenShift then if required it can wait until they complete etc. 

e.g. if a customer already has existing builds in TeamCity/Bamboo we can use a Jenkinsfile to orchestrate those builds then test & deploy the build artefacts in OpenShift

**Comment by jimmidyson** [comment]
I'd say it might be best to get Jenkins integration nice & tight first, bringing TeamCity/Bamboo into the picture later if there's a call for it.

**Comment by mfojtik** [comment]
@jimmidyson @jstrachan can you guys update this card? https://trello.com/c/ZFMTmd9d/413-5-autoconfigure-jenkins-build-jobs

**Comment by smarterclayton** [comment]
The big question seems to be should the Jenkins integration ALWAYS write back build objects that reflect jenkins builds, or simply write status back to the build config.  I *think* it would be better if Jenkins was writing things to build config like:


(just an example)
```
{
  ...
  "status": {
   "currentBuild": 5,
   "buildStatus": "inprogress", 
   "started": "324938403284",
  }
}
```

If Jenkins keeps that up to date (periodically writing status) then the OpenShift UI would be able to react to it.  We've talked about having the openshift builds be reflected back on the build config (we're doing that for deployments) so that it's easier to understand what is going on.

I would like to be able to provision a single jenkins per project and do single sign on via OAuth with the console - that's a couple of steps down the road though.  We could absolutely do the simplest possible integration today which is when a user creates the first build config with "Jenkins" we spin up an instance automatically (on our side) that uses your images.

**Comment by smarterclayton** [comment]
> We could absolutely do the simplest possible integration today which is when a user creates the first build config with "Jenkins" we spin up an instance automatically (on our side) that uses your images.

And we could build that in a way that informs service claims and helps us do a different integration.

I agree with the general concerns about being Jenkins specific - I guess it depends on how generic the user's request is - we could have it be ExternalCI and have a field where we could set the Jenkinsfile into it - the integrations could detect that it's Jenkinsfile type and "claim" it for Jenkins.  Maybe if we mock up what a flow would be where it's

1. new-app detects Jenkinsfile
2. creates build config with a type for now of Custom build config (for prototyping) and populates an env var with the Jenkinsfile
3. we create something that spins up a Jenkins instance from a template you guys provide in the namespace automatically if one doesn't exist
4. That jenkins instance has a service account that has credentials for you to talk to the API
5. You guys have a polling / watch loop that sees the Custom build config with the special env var
6. You then create a build in Jenkins, initialize all the magic, then write back some metadata to an annotation on the build config
7. When the build is done you write status back to the build config
8. OpenShift UI then displays the build config and sees your annotations, displays the links / starts up the jenkins extension to embed / link to / show your UI

If we could show that path, we can then figure out what the env / inputs / annotations / status writes are and we can then formalize it into the build config as an API change for the second iteration (productization).

**Comment by smarterclayton** [comment]
Does a 1=1 correspondence between an openshift build config an a Jenkins/Pipeline make sense?

**Comment by jstrachan** [comment]
@jimmidyson totally agreed - lets nail awesome Jenkins support from an implementation perspective first before worrying about anything else.

My comment was more to avoid calling everything a `JenkinsBuildStrategy` in the BuildStrategy type; maybe we call it a CIBuildStrategy or something so that its specifically defined to talk to some CI service via a named service at a given relative URL, with optional secrets etc. Then later on folks can try using it for Teamcity / Bamboo too etc.

**Comment by jstrachan** [comment]
@smarterclayton yes, there's a 1=1 correspondence between an openshift BuildConfig and a Jenkins job and a Jenkinsfile when using the Jenkins Pipeline. Then each build of the jenkins job maps to an openshift Build

**Comment by arilivigni** [comment]
FYI @smarterclayton you can also do this today with Jenkins Job Builder.  1 file to 1 multi-job.  I am working on support for JJB and the workflow plugin.  The format of JJB IMHO is a lot cleaner then a Jenkinsfile.  It is YAML based.

**Comment by smarterclayton** [comment]
I'm fine with that - I'm most concerned that you have the necessary
API to store status data in this model, and what bits of job config we
want to have in Openshift (to allow an API consumer to spawn types).
Ie maybe under Pipeline there is a type "pull-request", "approval",
"stage-prod" that is the template of the outcome job.  That enables

    oc new-build --pipeline=approval

Down the road

**Comment by smarterclayton** [comment]
Ben and I chatted yesterday - do we always need to create an Openshift
build?  Or just when you need a particular image built?  If a user
creates the build config, but you guys create and drive the builds
(and we make sure there is flow back).  I'd definitely like to make
the flow of user from creation into the Jenkins experience be seamless
- they don't necessarily need to be able to oc get builds unless we
want them to.

**Comment by jstrachan** [comment]
@smarterclayton a Jenkins Pipeline job is a single Build (in the OpenShift sense) that starts and keeps running until completion/failure and has output. 

Within the Jenkinsfile you can define stages which then are used to visualise its progress and group steps in long running pipelines (e.g. "make image", "run system tests", "approve promotion to Staging"). A particular Jenkinsfile may create a PR and wait for it to get merged/closed, or perform human approval or staging etc. We can then visualise the progress of the stages:
<img width="1112" alt="pipeline-stages" src="https://cloud.githubusercontent.com/assets/30140/12849909/f478b330-cc19-11e5-82f8-6cad9674959d.png">

Once a build is started; generally your only option is to cancel it or wait for it to complete. Depending on the human approval engine / mechanism you wish to use on your project, you may want to come up with some 'oc approve' command to approve a build that is waiting on human approval. You can use a Promote button in the fabric8 console today; or we can use +1 on gerrit as a voting mechanism. Ultimately human approval is maybe done better with a human workflow type task list; e.g. using jBPM which allows team based approvals with escalation etc. Other customers may prefer to use their issue tracker for approvals.

In terms of creating an OpenShift Build; if we create an OpenShift BuildConfig for a Jenkins build; then whenever we trigger the build via `oc start-build` we should create an OpenShift Build entity. It would be nice if the Jenkins build gets triggered (e.g. via a webhook or timer) that we also automatically create a Build too; so that we have a consistent view of these builds inside the OpenShift command line, REST API and console. But I guess we could make this configurable? Its not mandatory that there is an OpenShift Build for every Jenkins Pipeline job build - it does feel like always creating a Build would make things more consistent though

**Comment by smarterclayton** [comment]
If they're mostly for the use of Jenkins (to have the environment for S2I, etc), then there's a slight disincentive to show them.  If end users can interact, there's a benefit.  Ultimately, Jenkins is keeping them in sync - as a first step, having Jenkins keep the build config in sync is enough.

Also, start-build should be using buildconfig/instantiate, which for a Pipeline strategy should be bypassing creation of the openshift build object unless we really need it.  I think we would instead update the build config status / spec with the updated desired outcome.  And Jenkins would reconcile that and launch a new build.

**Comment by jstrachan** [comment]
creating an OpenShift Build for a jenkins job build is more for the user to see whats going on inside jenkins without having to leave the oc command line or the OpenShift console. Jenkins doesn't care either way; so maybe its a configuration option to whether or not we create the Build objects in OpenShift for the Jenkins job builds?

**Comment by smarterclayton** [comment]
Maybe - I definitely want to see the technical plan for how we integrate
s2i and the jenkins work - it's not clear to me now what benefit an s2i
user will get if their builds run in a different context, so i would expect
at least some of the time jenkins to be creating an s2i build.  Maybe we
need to sort that out first and then that'll tell us what the right thing
is here.

But in general, it sounds like we're in agreement that Jenkins is going to
be a per project (with some exceptions, maybe for very small clusters)
"Jenkins build controller", which acts to react to openshift API requests
and update status on API objects (starting small), with UI integration that
means we need some form of single sign on, and some way of ensuring that we
have good credential sharing.  It's a controller on steroids (has a UI and
its own API) but that's fine - we want that.

Jordan and I were talking about the security angle - does fabric today
include an oauth integration from jenkins that can do the oauth dance to a
provider for single sign on?  We aren't going to be able to give Jenkins
the user's token in the long run, so we need to queue up the "how does the
user single sign on with a project jenkins that doesn't get access to act
as them outside of the project".  I'd rather not have one-off users - I'd
like something to be really clean there.  We are *thinking* that the
jenkins service account can act as an oauth client (it has a client
password, it's secret), and so if we have per namespace grants an oauth
flow from openshift -> jenkins -> openshift would look like:

  * jenkins is an oauth client registered with openshift implicitly per
namespace
  * user browses to jenkins in console
  * jenkins does the oauth dance
  * openshift shows a "do you want to let jenkins know who you are"
  * user accepts
  * jenkins gets a user identity token which it can check ACL to the
console against
  * user is signed in to jenkins

We can make the flow auto granting probably - so you don't have to answer
the question normally.

**Comment by jstrachan** [comment]
@smarterclayton a Jenkins pipeline build may invoke one or more S2I builds; then may orchestrate testing them in another environment, then promoting them. Or the Jenkins pipeline may built its own images using the new binary builder or using raw docker or trigger existing Jenkins or external CI jobs to build the images.  

If someone wants S2I and Pipeline together in a single BuildConfig; we could look later on at how we can avoid having 2 BuildConfigs; one for the pipeline and one for the S2I (as the configuration of the S2I BuildConfig could be inlined inside the Jenkinsfile) so that its easier to setup projects. 

Though right now, AFAIK the [openshift jenkins pipeline plugin](https://github.com/jenkinsci/openshift-pipeline-plugin) can only trigger an existing S2I BuildConfig so `oc new-app` would need to be told to create 2 builds; one for the pipeline and one for the S2I build then we'd refer to the S2I BuildConfig inside the Jenkinsfile. Then there's the issue that you can run the S2I BuildConfig on its own without running the rest of the pipeline which sounds wrong; you should really only be able to run the 1 Pipeline really. 

So I kinda prefer the simplicity of invoking the S2I build capabilities inline in the Pipeline without requiring a separate BuildConfig for S2I; so there's less noise or complexity. We just need to embed into the Jenkins Pipeline build steps the ability to run the S2I builder; then things are nice and simple - there's 1 BuildConfig that does the S2I stuff inside the pipeline and can orchestrate any pre and post steps.

I agree on the jenkins per project approach as the build orchestrator. 

We've not done an OAuth plugin for Jenkins yet; we're just running it with its own ServiceAccount.

**Comment by rawlingsj** [comment]
FWIW there's an couple of issues raised on the openshift jenkins plugin that relate to adding general s2i build capabilities to Jenkins pipeline 
https://github.com/openshift/jenkins-plugin/issues/7#issuecomment-163944581

And another which is focused on promoting images across projects / namespaces https://github.com/openshift/jenkins-plugin/issues/8

**Comment by smarterclayton** [comment]
Ben and I did a quick touch base yesterday - one other topic that has been sloshing around that is relevant here is the idea of different kinds of Jenkins:

1. Simple stateless "just do the basics for me" - a jenkins that needs no persistent storage and is entirely driven on the state in OpenShift
2. Simple "almost stateless" that can store jenkins config back in OpenShift as a config, or in a Git repo that is the input for a layered build of jenkins (probably the former, allows us to rapidly update) - this jenkins can be customized but doesn't need anything on disk
3. Mutable jenkins - has a persistent disk, survives longer, users get in and really muck around with

I kind of break these into "how much do I need **jenkins**" vs "how much do I need a **pipeline**".  You might start on #1 by just creating the jenkins build config - we spin up a stateless jenkins that uses the config from the jenkins build config.  Then you decide you want to enable some plugins, and start customizing other tools (plugins are in the image, but config has to live somewhere).  So the config gets written back to a configmap for jenkins stored in etcd (which is ok for up to 100k or so).  But then you really want to preserve builds, artifacts, start doing more sophisticated things, so you attach a PVC and it then holds your data.

The first two are aspirational, but one of the advantages of thinking deliberately about modes where that is in play is that makes the control point for versioning and the template definition under the admins or user's control (no, that jenkins broke me, rollback to version X and lock there).  It means that for the large class of simple jenkins users, they don't need their own PV, and if they have their own Jenkinsfile in their git repo they don't even really need to edit / manage from Jenkins.  That class of users can also be easily migrated centrally (we roll out a fix to the template pipeline they are using, or security update, etc).  Then when you're ready (big organizational user, running builds for a big project or team) you can throw a PV on and become an authoritative Jenkins server.  Those users probably are going to fine tune jenkins anyway - but because they're just a few of them per org you get the nice tradeoff that simple users don't have to manage jenkins but hardcore users can.

Done right, that makes the demo flows all that more impressive - it makes Jenkins seem like a genie at the beck and call of the user, and you can flip that switch from "demo / throwaway" to "personal build farm"

**Comment by jstrachan** [comment]
Agreed. I'd maybe split things into whether or not folks are using a `Jenkinsfile` or not rather than the `pipeline` distinction. You can use a Jenkinsfile for a regular CI build too really.

If folks are not using a Jenkinsfile then those options all make sense - e.g. creating arbitrary persistent jobs, adding plugins, tools and custom configurations etc.

If folks are using a Jenkinsfile then there's not really any persistent storage required in Jenkins; there's no need to muck with the Jenkins installation to add plugins, install tools or configure plugins/jobs or whatever - everything is in the Jenkinsfile. You can trash Jenkins at any time, recreate it and point it at your github organisation and Jenkins will automatically create all the jobs for you.

Then if a BuildConfig is using Jenkins, we can automate the creation of Jenkins jobs for the Jenkinsfiles in a similar way; we don't have to replicate job state from Jenkins into ConfigMaps or anything. So a stateless Jenkins would do the trick really for Jenkinsfiles; whether you are using them as vanilla CI or as pipelines etc.

I'd like us to be opinionated in OpenShift 3.3 for the use of Jenkins + Jenkinsfiles to encourage customers to put everything in the Jenkinsfile, to version it automatically in source code - and use Jenkins in the 2.x way - using docker images for all the tools they need (to avoid needing Jenkins plugins or installing tools) and avoids the need for PVs. 

This also avoids a whole set of problems where a developer wants a new version of some plugin or tool and has to wait for some admin to update Jenkins or recreate a new image etc. 

We should also encourage folks to use  [kubernetes pipelines](https://github.com/fabric8io/kubernetes-workflow) so that folks can reuse [secrets](https://github.com/fabric8io/kubernetes-workflow/blob/master/kubernetes-steps/readme.md#using-secrets) within Jenkinsfiles (e.g. mounting ~/.m2/settings.xml files for maven with login/pwd to access docker/nexus/artifactory repos) to avoid most of the need for custom jenkins images or installations. Previously a really common reason for custom jenkins images was secrets.

Finally the awesome thing about kubernetes is; if folks want to replace the OOTB Jenkins image with their own custom image with different configuration / plugins they can. Though with Jenkinsfiles that should not really be needed; but folks can remove the batteries if they wish; either using their own image/template or service linking to an external Jenkins installation.

**Comment by jstrachan** [comment]
BTW I'd prefer us to prioritise using Jenkins + Jenkinsfiles properly with docker / kubernetes inside OpenShift first and really nail it from a developer experience perspective of how best to use OpenShift + Jenkins nicely together before we get too lost in the weeds dealing with legacy approaches to Jenkins (e.g. replicating Jenkins 1.x jobs, plugins and configuration into ConfigMaps and so forth).

**Comment by bparees** [comment]
now that we're actually doing this (and have the cards to reflect the feature work), closing this.
